src/mlpack/methods/ann/ffn.hpp

```diff
@@ -522,23 +522,6 @@ class FFN
 } // namespace ann
 } // namespace mlpack
 
-//! Set the serialization version of the FFN class.  Multiple template arguments
-//! makes this ugly...
-namespace boost {
-namespace serialization {
-
-template<typename OutputLayerType,
-         typename InitializationRuleType,
-         typename... CustomLayer>
-struct version<
-    mlpack::ann::FFN<OutputLayerType, InitializationRuleType, CustomLayer...>>
-{
-  BOOST_STATIC_CONSTANT(int, value = 2);
-};
-
-} // namespace serialization
-} // namespace boost
```

Comment:
I have reset the version in all mlpack codebase. I think it will be easier to activate that as we change the classes in the future? I am not sure, it is your call.

Reply:
Yeah; I think we have to "prepare" for that.  Consider this serialize function:

```
class C
{
  int x;
  int y;
  int z;

  void serialize(Archive& ar)
  {
    ar(x);
    ar(y);
    ar(z);
  }
};
```

But if we ever want to change that and add another parameter `a` to the definition of `C`, now we break reverse compatibility.  There's no way to write a reverse-compatible `serialize()` function.  So instead I would suggest an implementation like this:

```
class C
{
  int x;
  int y;
  int z;

  void serialize(Archive& ar)
  {
    int version = 1;
    ar(version);

    ar(x);
    ar(y);
    ar(z);
  }
};
```

And then, if we wanted to add `a`, we could do it in a reverse-compatible way:

```
class C
{
  int x;
  int y;
  int z;
  int a;

  void serialize(Archive& ar)
  {
    int version = 2;
    ar(version); // If an old model is loaded, version will be set to 1.

    ar(x);
    ar(y);
    ar(z);
    if (version >= 2)
      ar(a);
    else
      a = 0.0; // Set a default value for `a` since it was not serialized in the old version.
  }
};
```

So, this means we'll have to add that `int version; ar(version);` snippet to basically every mlpack class that we want to be reverse-compatible.  I would say it's okay to do that for just about every class, although we may want to avoid it for things we hold huge vectors of.  For instance, maybe we want to avoid it for most classes in `core/`, and instead just have versions for things in `methods/`.  (Also, maybe we want a `uint8_t` for `version` to save 3 bytes.)